ext/drx/drx_time_scale.c

```diff
@@ -216,6 +221,40 @@ inflate_timespec(void *drcontext, struct timespec *spec, int scale,
            dr_get_thread_id(drcontext), scale, spec->tv_sec, spec->tv_nsec);
 }
 
+static bool
+inflate_absolute_timespec(void *drcontext, struct timespec *spec, int scale,
+                          drx_time_scale_type_t type, bool realtime)
+{
+    /* First, convert to relative. */
+    struct timespec cur_time;
+    int res =
+        dr_invoke_syscall_as_app(drcontext, SYS_clock_gettime, 2,
+                                 realtime ? CLOCK_REALTIME : CLOCK_MONOTONIC, &cur_time);
+    if (res != 0) {
+        NOTIFY(0, "Failed to call clock_gettime\n");
+        increment_attempt_and_failure(type);
+        return false;
+    }
+    int64_t cur_nanos = (int64_t)cur_time.tv_sec * MAX_TV_NSEC + cur_time.tv_nsec;
+    int64_t target_nanos = (int64_t)spec->tv_sec * MAX_TV_NSEC + spec->tv_nsec;
+    int64_t diff_nanos = target_nanos < cur_nanos ? 0 : target_nanos - cur_nanos;
+    if (diff_nanos == 0) {
+        diff_nanos = ZERO_PRE_INFLATE_NSEC;
+        increment_convert_zero(type);
+    }
+    struct timespec rel;
+    rel.tv_sec = diff_nanos / MAX_TV_NSEC;
+    rel.tv_nsec = diff_nanos % MAX_TV_NSEC;
+    /* Now inflate the relative. */
+    inflate_timespec(drcontext, &rel, scale, type);
+    /* Now convert back to absolute. */
+    int64_t scaled_diff_nanos = (int64_t)rel.tv_sec * MAX_TV_NSEC + rel.tv_nsec;
+    int64_t scaled_target_nanos = cur_nanos + scaled_diff_nanos;
+    spec->tv_sec = scaled_target_nanos / MAX_TV_NSEC;
```

Comment:
Would it be useful to define macros for to_nanos and from_nanos?

Reply:
Done.